SimplifiedTests/NYPLContentDownloaderTests.swift

```diff
@@ -0,0 +1,156 @@
+        case .success:
+          expectation.fulfill()
+          if !["1", "2", "3"].contains(url.lastPathComponent) {
+            XCTFail()
+          }
+        case .failure:
+          expectation.fulfill()
+        }
+      }
+    }
+    
+    wait(for: [expectation], timeout: 5)
+  }
+  
+  /// When downloading multiple items, If none of the download fails, NYPLContentDownloader should not
+  /// stop until all the items are downloaded.
+  func testAllTasksShouldExecuteIfNoFailureOccurs() {
+    let expectation = XCTestExpectation(description: "If no failure occurs, all requests should be executed!")
+    expectation.expectedFulfillmentCount = allSucceedingURLs.count
+    
+    for url in allSucceedingURLs {
+      contentDownloader.downloadItem(from: url) { (result) in
+        switch result {
+        case .success:
+          expectation.fulfill()
+        case .failure:
+          XCTFail()
+        }
+      }
+    }
+
+    wait(for: [expectation], timeout: 5)
+  }
+  
+  /// NYPLContentDownloader and its network executor should deinitialize successfully.
+  func testNetworkExecutorAndContentDownloaderShouldDeinitUponCompletion() {
+    let expectation = XCTestExpectation(
+      description: "NYPLAxisContentDownloader and NYPLAxisNetworkExecutor should deinitialize upon completion")
+    expectation.expectedFulfillmentCount = 2
+    
+    customContentDownloader?.deinitialzed = {
+      expectation.fulfill()
+    }
+    
+    if let executor = customContentDownloader?.networkExecuter as? MockAxisNetworkExecutor {
+      executor.deinitialzed = {
+        expectation.fulfill()
+      }
+    }
+    
+    for url in allSucceedingURLs {
+      customContentDownloader?.downloadItem(from: url) { (_) in }
+    }
+    
+    customContentDownloader = nil
+    
+    wait(for: [expectation], timeout: 10)
+  }
+  
+  /// NYPLContentDownloader and its network executor should deinitialize successfully.
+  func testNetworkExecutorAndContentDownloaderShouldDeinitUponFailure() {
+    let expectation = XCTestExpectation(
+      description: "NYPLAxisContentDownloader and NYPLAxisNetworkExecutor should deinitialize upon failure")
+    expectation.expectedFulfillmentCount = 2
+    
+    customContentDownloader?.deinitialzed = {
+      expectation.fulfill()
+    }
+    
+    if let executor = customContentDownloader?.networkExecuter as? MockAxisNetworkExecutor {
+      executor.deinitialzed = {
+        expectation.fulfill()
+      }
+    }
+    
+    for url in urlsWithOneFailure {
+      customContentDownloader?.downloadItem(from: url) { (_) in }
+    }
+    
+    customContentDownloader = nil
+    
+    wait(for: [expectation], timeout: 10)
+  }
+
+}
+
+private class MockAxisNetworkExecutor: NYPLAxisNetworkExecuting {
+  
+  var deinitialzed: (() -> Void)?
+  
+  func GET(_ request: URLRequest, completion: @escaping (Result<Data, Error>) -> Void) -> URLSessionDataTask {
+    let task = URLSession.shared.dataTask(with: request)
```

Comment:
I'm struggling to understand this part. This `task` is intentionally not resumed, correct? however you are returning it to the caller, which is the `NYPLContentDownloader` instance at the top of this file, and I think that does actually execute it. Am I reading this right?

I was going to say that it is preferable to avoid hitting the internet in unit tests, but I'm not sure we actually do that here. But if we don't, I don't quite understand the reason for the Timer.

The approach I've taken in the past for mocking server / networking behavior is something like in the NYPLNetworkExecutorMock, but I'm not entirely sure that applies here.

Reply:
Now that I think about it, if I have to explain why I did something in a certain way it means I did not do it well, and if we merge that code then maintaining it will be harder than it has to be. Switched to using NYPLNetworkExecutorMock.